CUDA support in the CSR layout: constructors

aocsa opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57403 CUDA support in the CSR layout:  CUDA addmm
* #57275 CUDA support in the CSR layout: sparse_to_dense/add_sparse_csr
* **#57274 CUDA support in the CSR layout: constructors**

closes gh-56485, gh-57211 and  gh-58762

Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

aocsa committed:
CUDA support in the CSR layout: constructors

[ghstack-poisoned]

ezyang commented on test/expect/TestSparseCSRCPU.test_sparse_csr_print_cpu.expect:
```diff
@@ -96,22 +99,25 @@ tensor([ 0.0548,  0.2650, -0.8181, -0.5354,  0.4537, -0.7625, -0.2098,  0.4398,
-                            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-                            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
-                            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]),
+tensor(crow_indices=tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 10, 10,
```
huh, was the old printing logic wrong? :)

aocsa committed:
Update on "CUDA support in the CSR layout: constructors"


Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: constructors"


Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

mruberry commented:
Looks like @ezyang and @pearu are reviewing this stack, but let me know if I should look at something

pearu reviewed (changes requested):
Thanks, @aocsa for implementing CUDA support to the CSR layout! In my review I have pointed out the following concerns:
- in device check matches, should one use `get_device()` or `options.backend()` approach? But definitely not both, imho.
- currently, a mix of devices is allowed for certain combinations of crow_indices/col_indices/values. I think we should restrict these to a simple rule: either allow all device combinations or just a single one.
- some checks copy the crow_indices to CPU while only the first and the last element is needed.
- nits: duplications of checks, creating csr from list inputs is not tested on CPU

ezyang commented:
ok gonna wait for pearu's comments to be addressed

aocsa commented on test/test_sparse_csr.py:
```diff
@@ -77,7 +75,47 @@ def test_factory_size_check(self, device, dtype):
+    @dtypes(*torch.testing.get_all_dtypes(include_bool=False, include_half=False,
+                                          include_bfloat16=False, include_complex=False))
+    def test_factory_device_type_inference(self, device, dtype):
+        sparse = torch.sparse_csr_tensor([0, 2, 4],
+                                         [0, 1, 0, 1],
+                                         [1, 2, 3, 4],
+                                         size=(2, 10),
+                                         dtype=dtype,
+                                         device=device)
+
+        self.assertEqual((2, 10), sparse.shape)
+        self.assertEqual(torch.tensor([0, 2, 4], dtype=torch.int32, device=device), sparse.crow_indices())
+        self.assertEqual(torch.tensor([0, 1, 0, 1], dtype=torch.int32, device=device), sparse.col_indices())
+        self.assertEqual(torch.tensor([1, 2, 3, 4], dtype=dtype, device=device), sparse.values())
```
It runs with `device=='cpu'` as well as you can see here.
https://github.com/pytorch/pytorch/blob/6c6184f3d38fae4756cfbb2fa2cdfb4ae3db06c4/test/test_sparse_csr.py#L96
 It generates a different kind of combinations among ('cpu', 'cuda', None) to test device type inference.

pearu replied:
See https://github.com/pytorch/pytorch/blob/6c6184f3d38fae4756cfbb2fa2cdfb4ae3db06c4/test/test_sparse_csr.py#L78 doesn't this disable CPU runs?

aocsa replied:
It doesn't disable CPU. If @onlyCUDA is enabled, the test was designed to work with CUDA because the functions you are using was compiled with CUDA, so that means that it was compiled for CPU as well.  To probe that I run the test with @onlyCUDA  enabled and @onlyCPU enabled.   So basically it uses onlyCUDA to not run twice.

pearu replied:
I can see the point of avoiding running the tests twice. However, when pytorch is built without CUDA support, the tests are not run, AFAIRFC.

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28112689).

aocsa commented on aten/src/ATen/SparseCsrTensorImpl.cpp:
```diff
@@ -139,6 +139,14 @@ void SparseCsrTensorImpl::set_member_tensors(
       values.get_device(),
       ")");
 
+  TORCH_CHECK(
+        crow_indices.options().backend() == values.options().backend(),
```
> Add
> 
> ```
> TORCH_CHECK(
>         col_indices.options().backend() == values.options().backend(),..
> ```
> 
> for consistency.
> 
> In fact, the check
> 
> ```
> TORCH_CHECK(
>       values.is_cuda() || col_indices.get_device() == crow_indices.get_device(),...
> ```
> 
> is confusing: if `values` uses cuda, then `col_indices` and `crow_indices` can be stored in different devices. Why is that?

I think this condition is old and wrong, I will remove this part `values.is_cuda()`.  So adding
`TORCH_CHECK(col_indices.get_device() == values.backend(),`  is not needed anymore as previous assert checks `crow_indices.get_device() == col_indices.get_device()`. 

> 
> Also, what is the difference between
> 
> ```
> crow_indices.options().backend() == values.options().backend()
> ```
> 
> and
> 
> ```
> crow_indices.get_device() == values.get_device()
> ```

`.backend()` could include CPU, CUDA but also SparseCPU or SparseCsrCPU. `.get_device`  could include CPU, CUDA but also META (for tensors with no data). I will use `.get_device` for consistency.

> 
> ? If none, use either `backend()` or `get_device()`, but not both.

Ok

aocsa commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -103,15 +109,14 @@ Tensor sparse_csr_tensor(
-      auto crow_indices_accessor = crow_indices.accessor<index_t, 1>();
+      auto cpu_crow_indices = crow_indices.to(kCPU);
+      auto crow_indices_accessor = cpu_crow_indices.accessor<index_t, 1>();
       TORCH_CHECK(
```
Is not duplicated, as these checks are in different sparse_csr_tensor factory functions. The first one with size included the second one without including size.

pearu replied:
Ah, right, the github diff view is doing tricks for me. So, nvm.

aocsa commented on test/test_sparse_csr.py:
```diff
@@ -77,7 +75,47 @@ def test_factory_size_check(self, device, dtype):
+
+        cpu_cuda = ('cpu', 'cuda')
+        cpu_cuda_none = cpu_cuda + (None,)
+        for indices_device, values_device, device in itertools.product(cpu_cuda,
```
Ok

aocsa commented:
> Thanks, @aocsa for implementing CUDA support to the CSR layout! In my review I have pointed out the following concerns:
> 
> * in device check matches, should one use `get_device()` or `options.backend()` approach? But definitely not both, imho.
> * currently, a mix of devices is allowed for certain combinations of crow_indices/col_indices/values. I think we should restrict these to a simple rule: either allow all device combinations or just a single one.
> * some checks copy the crow_indices to CPU while only the first and the last element is needed.
> * nits: duplications of checks, creating csr from list inputs is not tested on CPU

Thanks, @pearu, for the review. I rebased this PR on top of the latest changes, polished a few things based on the feedback. Let me know your thoughts.

aocsa committed:
Update on "CUDA support in the CSR layout: constructors"


Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: constructors"


Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

aocsa committed:
Update on "CUDA support in the CSR layout: constructors"


Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

pearu commented on aten/src/ATen/SparseCsrTensorImpl.cpp:
```diff
@@ -129,15 +129,22 @@ void SparseCsrTensorImpl::set_member_tensors(
+
+  TORCH_CHECK(
+        crow_indices.get_device() == values.get_device(),
+        "device of csr indices (",
```
nit: `"device of crow_indices"`

pearu reviewed (changes requested):
LGTM, just a minor nit and two unresolved comments:
- [construction from lists is not tested when using cuda-disabled build]( https://github.com/pytorch/pytorch/pull/57274/files/6c6184f3d38fae4756cfbb2fa2cdfb4ae3db06c4#r624748940)
- [independent device tests for crow_indices and col_indices](https://github.com/pytorch/pytorch/pull/57274/files/6c6184f3d38fae4756cfbb2fa2cdfb4ae3db06c4#r624946232)

aocsa commented:
> LGTM, just a minor nit and two unresolved comments:
> 
> * [construction from lists is not tested when using cuda-disabled build](https://github.com/pytorch/pytorch/pull/57274/files/6c6184f3d38fae4756cfbb2fa2cdfb4ae3db06c4#r624748940)

I moved these to test_sparse_csr_constructor which test both CPU/CUDA

> * [independent device tests for crow_indices and col_indices](https://github.com/pytorch/pytorch/pull/57274/files/6c6184f3d38fae4756cfbb2fa2cdfb4ae3db06c4#r624946232)

I had these changes before and lost them, thanks for notice this. 

Thanks @pearu.

aocsa committed:
Update on "CUDA support in the CSR layout: constructors"


Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

pearu reviewed (approved):
LGTM

aocsa committed:
Update on "CUDA support in the CSR layout: constructors"


Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28112689).

ezyang commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -212,7 +212,7 @@ Tensor internal_new_from_data(
   }
 
-  if (THPVariable_Check(data)) {
+  if (PyObject_IsInstance(data, THPVariableClass) == 1) {
```
What's going on here? This wasn't here on my initial review

aocsa replied:
Some test constructions were not considered in the first commit.  I am talking about  `sparse_csr_tensor` construction from lists. 
https://github.com/pytorch/pytorch/blob/b07d38a05df0ef7cf313b7751e86bce97692778d/test/test_sparse_csr.py#L36
Something weird was happening when I was trying to create these tensors, the function    `THPVariable_Check/PyObject_IsInstance` returned `-1` (an error) or `0` (the expected one). This caused segmentation faults, I tracked every function call from the stack and there is no changes in the input. That's why I changed this condition to a more restricted one, I could find any other solution.

ezyang replied:
Hmm, I don't really understand this explanation. Why did returning -1/0 cause segmentation faults? Is it possible to print out what `data` is in the stack traces that caused the segfault?

Given that the line of code you quoted is `sparse_csr_tensor`, what seems more likely to me is that there is insufficient checking in the `sparse_csr_tensor` binding for, e.g., lists. If that's the case, you should fix it there, not here... (unless you can give a good argument for why this line of code is wrong)

aocsa replied:
I figured out the error, the way how processing input parameters at `sparse_csr_tensor_ctor` functions were wrong. This change was reverted.

ezyang commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -47,18 +47,24 @@ Tensor sparse_csr_tensor(
   AT_DISPATCH_INDEX_TYPES(crow_indices.scalar_type(), "csr_construct_check", [&] {
-    auto crow_indices_accessor = crow_indices.accessor<index_t, 1>();
+    index_t first_index_value = crow_indices.select(0, 0).item<index_t>();
+    index_t last_index_value = crow_indices.select(0, crow_indices.numel() - 1).item<index_t>();
```
If we're assuming `crow_indices` is contiguous (which I'm pretty sure we should), this code is worse because the `select` operation is relatively heavyweight and has to allocate a whole new TensorImpl before you just poke out a single item. The accessor code was better.

aocsa replied:
One approach to support for CUDA requires to have `crow_indices` in CPU (a copy operations CUDA->CPU) to check invariants. @pearu did some quick tests and he figured out that just getting one element is less expensive than  copying to cpu the whole Tensor.
    
https://github.com/pytorch/pytorch/pull/57274#discussion_r624956412
```
>>> start=time.time();b1=a.to(device='cpu'); b=b[0], b[-1]; print(time.time()-start)
0.01662921905517578
>>> start=time.time();b=tuple(a[::a.numel()-1].to(device='cpu')); print(time.time()-start)
0.0002875328063964844
>>> start=time.time();b=a[0].to(device='cpu'),a[-1].to(device='cpu'); print(time.time()-start)
0.013242721557617188
```
What I think is that I can create two paths to do that, the simple one using `accessor` for CPU device and another for CUDA device I will check times using this approach.

ezyang replied:
That is fair. Actually I think this is inextricably linked to the discussion at https://github.com/pytorch/pytorch/pull/57274#discussion_r626829673 so maybe we will know what to do here after we resolve that thread.

aocsa replied:
Solved here: 
https://github.com/pytorch/pytorch/pull/57274#discussion_r626829673

ezyang commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -103,20 +109,19 @@ Tensor sparse_csr_tensor(
   if (col_indices.numel() > 0) {
     size[0] = crow_indices.numel() - 1;
-    Tensor max_col_indices = std::get<0>(col_indices.max(0, false));
+    Tensor max_col_indices = std::get<0>(col_indices.max(0, false)).to(kCPU);
```
SYNC!!! You sure about this?

aocsa replied:
Umhhh let me review more about CUDA SYNC. I know that I will learn something interesting in the process.

pearu replied:
This is what COO tensor constructor does as well.

When you ask this, @ezyang , the question arises: should we trust the value-content of `crow_indices` and `col_indices` (or `indices` in the case of COO) provided by users/callers, and not perform these value-based integrity checks?

If yes, then it will speed up the constructions of sparse tensors (especially those using cuda storage) but there is a risk that invalid data will crash python at some point. For example, crow_indices must be an increasing sequence and currently, we don't have a test for it. Here's an example that will crash python because of this:
```
torch.sparse_csr_tensor(torch.tensor([0, 40, 3]), torch.tensor([2, 0, 1]), [1, 2, 3]).to_dense()
```
(the crash happens in `to_dense` call and could be avoided there but it is not really `to_dense` fault that CSR tensor contains invalid indices data, IMHO).

If no, what would be the recommended way to perform value-based integrity checks when using CUDA storage?

Possible options:
1. have unsafe tensor constructors that skip the value-based integrity checks; to be used only when the caller ensures that integrity conditions are met
2. perform value-based integrity checks only when using CPU storage

ezyang replied:
(1) is typically a good idea, because when you are writing kernels you typically want to assume the integrity checks are upheld (otherwise the algorithm is just wrong) and if you force the integrity checks always you will effectively sync on every kernel call.

Another general principle which is useful to think about is that we are willing to pay some performance cost to do error checking in kernels, but usually this manifests as willingness to add CUDA asserts (e.g., when we do embedding lookups, we will CUDA assert that they are in bounds). CUDA asserts are less expensive than syncing and might be a good way of doing integrity checks in other situations.

cc @ngimel and @fmassa for other opinions

aocsa replied:
I added  an `unsafe csr tensor constructors` that skip the value-based integrity checks.
Still need to double-check indices invariants implementation.   See [CSR tensor invariants](https://github.com/pearu/pearu.github.io/blob/pearu-csr_invariants/csr_tensor_invariants.md)
 
cc @pearu

ezyang commented on test/test_sparse_csr.py:
```diff
@@ -29,10 +29,22 @@ def test_sparse_csr_constructor_shape_inference(self, device, dtype):
+        self.assertEqual((2, 2), sparse.shape)
+        self.assertEqual(torch.tensor([0, 2, 4], dtype=torch.int32, device=device), sparse.crow_indices())
+        self.assertEqual(torch.tensor([0, 1, 0, 1], dtype=torch.int32, device=device), sparse.col_indices())
+        self.assertEqual(torch.tensor([1, 2, 3, 4], dtype=dtype, device=device), sparse.values())
```
I suggest putting this in a separate test method

ezyang commented:
This PR has changed substantially and substantively since my original review, so unfortunately I have some more questions, sorry about the second round

aocsa committed:
addressed review comments on "CUDA support in the CSR layout: constructors"


closes gh-56485

Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

aocsa commented:
> This PR has changed substantially and substantively since my original review, so unfortunately I have some more questions, sorry about the second round

I addressed review comments, let me know your thoughts @ezyang, cc @pearu

pearu commented:
@aocsa @ezyang see [the state of CSR tensor invariants](https://github.com/pearu/pearu.github.io/blob/pearu-csr_invariants/csr_tensor_invariants.md) that reveals two issues:
- not all invariants are checked
- `sparse_csr_tensor(crow_indices, col_indices, values)` has many duplicated checks

In fact, there seem to be some opportunities to better engineer ensuring all the invariants.

@aocsa could you address the above in this PR?

ezyang commented:
Some other high level comments related to invariants: I imagine that a lot of our CUDA kernels will be coming from cuSparse (or from up and coming research libraries like https://github.com/google-research/sputnik ). It would be really helpful for prospective kernel writers if we could clearly indicate how our internal nomenclature (crow is nonstandard) maps to the cusparse documentation, for example. (Or better yet, rename our nomenclature to match the prevailing libraries). For example, cuSparse documents CSR as follows:

* nnz | (integer) | The number of nonzero elements in the matrix.
* csrValA | (pointer) | Points to the data array of length nnz that holds all nonzero values of A in row-major format.
* csrRowPtrA | (pointer) | Points to the integer array of length m+1 that holds indices into the arrays csrColIndA and csrValA. The first m entries of this array contain the indices of the first nonzero element in the ith row for i=i,...,m, while the last entry contains nnz+csrRowPtrA(0). In general, csrRowPtrA(0) is 0 or 1 for zero- and one-based indexing, respectively.
* csrColIndA | (pointer) | Points to the integer array of length nnz that contains the column indices of the corresponding elements in array csrValA.

Furthermore, https://docs.nvidia.com/cuda/cusparse/index.html#cusparse-generic-spvec-create specifies that NNZ is stored *on host*, so packing it into `crow_indices[nrows]` is bad because you won't be able to call into cuSparse kernels using this representation. Storing nnz on host may also solve some of your synchronization problems that I was complaining about earlier in review.

pearu commented:
> Furthermore, https://docs.nvidia.com/cuda/cusparse/index.html#cusparse-generic-spvec-create specifies that NNZ is stored _on host_, so packing it into `crow_indices[nrows]` is bad because you won't be able to call into cuSparse kernels using this representation. Storing nnz on host may also solve some of your synchronization problems that I was complaining about earlier in review.

@ezyang , no, NNZ is not packed to `crow_indices[nrows]`. The relation is another way around: the value of `crow_indices[nrows]` must be equal to NNZ. This is an implication from CSR format specification and not our implementation choice. Since `crow_indices` is provided by a user, we'll need to check that the invariant `crow_indices[nrows] == NNZ` holds (unless we choose to trust the user which I believe is not the case).

ngimel commented:
Users typically want to have unsafe nonsynchronizing versions of the operations, where they guarantee that the data is valid (and if it's not, their program is in error).

ezyang commented:
>  no, NNZ is not packed to crow_indices[nrows]. The relation is another way around: the value of crow_indices[nrows] must be equal to NNZ. This is an implication from CSR format specification and not our implementation choice.

Oops, this is my bad, I forgot about this, thank you for the clarification. (The high level point of making it clear how to map our fields to cuSparse nomenclature still stands though!)

pearu commented:
> (The high level point of making it clear how to map our fields to cuSparse nomenclature still stands though!)

Yes, that was also my plan. It is just too late here to give this immediately..

aocsa commented:
> @aocsa @ezyang see [the state of CSR tensor invariants](https://github.com/pearu/pearu.github.io/blob/pearu-csr_invariants/csr_tensor_invariants.md) that reveals two issues:
> 
> * not all invariants are checked
> * `sparse_csr_tensor(crow_indices, col_indices, values)` has many duplicated checks
> 
> In fact, there seem to be some opportunities to better engineer ensuring all the invariants.
> 
> @aocsa could you address the above in this PR?

Thanks for the review feedback comments. I rebased this PR, polished a few things based on the feedback, specifically:
-   Checks for CSR tensor invariants were reviewed and the missing ones were added. [See document](https://github.com/pearu/pearu.github.io/blob/pearu-csr_invariants/csr_tensor_invariants.md). 
     - Tests for these checks were added too 
     - Notes: 
         - Type invariants  `1.5 dtype is float32 (default), or float64, or int8, ..., or int64`, IMO is not needed as `AT_DISPATCH_TYPES` takes care about this.
         - `3.11 numel() == nrows * ncols`  is supported.  Now, `this->refresh_numel` is called everytime the `sizes` changes. 
         -  duplicated checks were removed
-  New unsafe tensor constructor `_sparse_csr_tensor_unsafe`  that skip the value-based integrity checks   
    - Test for this new constructor was added too  

IMO, still needs to double-check  value-based integrity checks (invariant 5.3). Let me know your thoughts  @pearu, cc @ezyang

aocsa committed:
review comments addressed on "CUDA support in the CSR layout: constructors"


closes gh-56485

Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

aocsa committed:
minor lint fix on "CUDA support in the CSR layout: constructors"


closes gh-56485

Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

pearu commented:
> Some other high level comments related to invariants: I imagine that a lot of our CUDA kernels will be coming from cuSparse (or from up and coming research libraries like https://github.com/google-research/sputnik ). It would be really helpful for prospective kernel writers if we could clearly indicate how our internal nomenclature (crow is nonstandard) maps to the cusparse documentation, for example. (Or better yet, rename our nomenclature to match the prevailing libraries).

@ezyang I have reviewed the existing naming conventions regarding the CSR format, see https://github.com/pearu/pearu.github.io/blob/pearu-csr_invariants/csr_naming_conventions.md
In the conclusions, I am satisfied with the current naming convention but I am also biased re the current choice. So, I am looking forward to proposals/suggestions for better naming conventions. However, I would not recommend using cuSparse naming convention without modifications because it is unpythonic in so many ways.. but nomenclature mapping sounds like a good idea.

What do you think?

pearu commented:
OT: Here's an idea for torch.sparse logo - flaming crow ;)
<img src="https://thumbs.dreamstime.com/z/illustration-fire-burning-crow-black-background-124168040.jpg" width=50%/>

pearu commented:
> Our naming is fine, just need a place in code where people can find it. (Table is cool too, but only if people can find it.)

OK, good.

Would a document such as `sparse_csr.rst` under https://github.com/pytorch/pytorch/tree/master/docs/source/notes/ be a good place where people could read about CSR details that are somewhat higher level than the details we have in API docs?

ezyang commented on aten/src/ATen/SparseCsrTensorImpl.cpp:
```diff
@@ -59,16 +60,12 @@ SparseCsrTensorImpl::SparseCsrTensorImpl(
-  auto empty_values = at::empty(nnz_size, values().options());
+  at::native::resize_output(crow_indices_, size[0] + 1);
+  at::native::resize_output(col_indices_, nnz_size);
+  at::native::resize_output(values_, nnz_size);
```
...should this have been stacked on https://github.com/pytorch/pytorch/pull/56744

pearu replied:
These `resize_output` calls are no-ops considering that `set_member_tensors` will reset the members:
https://github.com/pytorch/pytorch/blob/05a4b163a0deef2a9e9ba43594628e7346ae86c2/aten/src/ATen/SparseCsrTensorImpl.cpp#L134-L136

When creating a new CSR tensor in `_sparse_csr_tensor_unsafe`, can we eliminate the call to `resize_and_clear_` (for a new tensor there would be nothing to resize nor clear, and `set_members_tensors` will reset the members anyway) and just do the `sizes_and_strides_` and `numel` updates explicitly?

aocsa replied:
I agree with the idea to just do the sizes_and_strides_ and numel.  So I changed this and remove `resize_and_clear_`.

ezyang commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -8,12 +8,84 @@
+
+  // Indices invariants
+  AT_DISPATCH_INDEX_TYPES(crow_indices.scalar_type(), "csr_construct_check", [&] {
+    Tensor crow_indices_cpu = crow_indices.to(kCPU);
```
OK, so we're back to the slower version? (I was reasonably convinced by @pearu's comment that it is faster to only transfer the indices you needed, which I agree is true)

aocsa replied:
Previous to this update, only access to first and last element was required, however in order to completely cover the CSR invariants listed [here](https://github.com/pearu/pearu.github.io/blob/pearu-csr_invariants/csr_tensor_invariants.md):
> `5.3 crow_indices[i-1] <= crow_indices[i] for all i=1,...,nrows`

Access to the whole array is needed. We can go back to the  faster version only if this condition is skipped. cc @pearu

pearu replied:
I think we should keep testing 5.3 because:
- when the invariant 5.3 is not satisfied and we don't catch it, python will very likely crash when performing operations with CSR tensor with invalid `crow_indices`
- there is a way to skip expensive testing of invariants

pearu commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -8,12 +8,84 @@
+      "values must have dim=1 but got values.dim()=",
+      values.dim());
+  // Note, this check also enforces `crow_indices.numel() >= 1`
+  TORCH_CHECK(
+      crow_indices.numel() == (size[0] + 1),
+      "crow_indices.numel() must be size(0) + 1, but got: ",
+      crow_indices.numel());
+  TORCH_CHECK(
+      col_indices.size(0) == values.size(0),
```
Nit: If `t` is `crow_indices`, `col_indices`, or `values`, then we have `t.size(0) == t.numel()`. A quick benchmark shows that `t.numel()` is about 3-4x faster than `t.size(0)`. Although the timings are in the order of microseconds, I suggest using `numel()` instead of `size(0)` for consistency here and below.

pearu commented on torch/_tensor.py:
```diff
@@ -940,10 +940,14 @@ def to_sparse_csr(self):
-                                           size=coalesced_self.shape, dtype=coalesced_self.dtype)
+            device = coalesced_self.values().device
+            crow_indices = torch.tensor(ro, dtype=row_indices.dtype, device=device)
+            return torch.sparse_csr_tensor(crow_indices,
```
Since we can trust COO tensor correctness, we should be able to produce inputs such that unsafe CSR constructor could be used. Let's address this in the context of https://github.com/pytorch/pytorch/issues/57381 rather than here.

pearu commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -625,13 +618,13 @@ Tensor sparse_csr_tensor_ctor(c10::DispatchKey dispatch_key, at::ScalarType scal
                                            /*type_inference=*/type_inference);
     Tensor crow_indices =  internal_new_from_data(values.options(),
-      crow_indices_scalar_type, r.deviceOptional(DEVICE_TYPE_ARG), r.pyobject(CROW_INDICES_ARG),
+      kInt, r.deviceOptional(DEVICE_TYPE_ARG), r.pyobject(CROW_INDICES_ARG),
```
Does this mean that `crow_indices` and `col_indices` will be forced to have `int32` dtype? If so, it is a regression,

aocsa replied:
Actually for `crow_indices` and `col_indices` the data type will be inferred  based on the data type of the input (tensor, list or numpy array). In previous code there was a bug processing dtype, moreover  the `dtype` was forced to be int32 specifically for `list` inputs.

pearu commented on test/test_sparse_csr.py:
```diff
@@ -48,36 +47,198 @@ def test_sparse_csr_constructor(self, device, dtype):
+                                                size=(2, 10),
+                                                device=device)
+                    should_be_cuda = (device == 'cuda' or (device is None and values_device == 'cuda'))
+                    self.assertEqual(should_be_cuda, t.is_cuda)
```
Please add checks for the following conditions:
```
t.crow_indices().dtype == index_dtype
t.col_indices().dtype == index_dtype
t.values().dtype == dtype
t.crow_indices().device == t.values().device
t.col_indices().device == t.values().device
```

pearu commented on test/test_sparse_csr.py:
```diff
@@ -48,36 +47,198 @@ def test_sparse_csr_constructor(self, device, dtype):
+                                         dtype=dtype,
+                                         device=device)
+
+        self.assertEqual((2, 2), sparse.shape)
```
Add:
```
self.assertEqual(4, sparse.numel())
```

pearu commented on test/test_sparse_csr.py:
```diff
@@ -48,36 +47,198 @@ def test_sparse_csr_constructor(self, device, dtype):
+                                         dtype=dtype,
+                                         device=device)
+
+        self.assertEqual((2, 10), sparse.shape)
```
Add:
```
self.assertEqual(20, sparse.numel())
```

pearu reviewed (changes requested):
Thanks, @aocsa for the updates! The validation of CSR inputs and safe/unsafe construction of CSR tensors is much clearer now. Also, thanks for making `numel()` functional for CSR tensors.

I have some nits and suggestions for a couple of more checks (numel and dtype/device of member tensors), otherwise this PR looks good to me.

aocsa commented:
> Thanks, @aocsa for the updates! The validation of CSR inputs and safe/unsafe construction of CSR tensors is much clearer now. Also, thanks for making `numel()` functional for CSR tensors.
> 
> I have some nits and suggestions for a couple of more checks (numel and dtype/device of member tensors), otherwise this PR looks good to me.

Thanks Pearu, this week I'm on PTO I will work on this as soon as I've arrived.

aocsa committed:
addressed review comments - 2 on "CUDA support in the CSR layout: constructors"


closes gh-56485 , #57211

Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

bhosmer commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -651,18 +644,55 @@ Tensor sparse_csr_tensor_ctor(c10::DispatchKey dispatch_key, at::ScalarType scal
 }
 
+Tensor _sparse_csr_tensor_unsafe_ctor(c10::DispatchKey dispatch_key, at::ScalarType scalar_type, PyObject* args, PyObject* kwargs) {
+  TORCH_INTERNAL_ASSERT(!isSparseCsr(dispatchKeyToBackend(dispatch_key)));
```
Probably want to check `!isSparse()` here too, right? (And similarly add `!isSparseCsr()` to the COO version)

aocsa replied:
Sure, just to be sure that everything is ok these checks are added too.

pearu replied:
> Probably want to check `!isSparse()` here too, right? (And similarly add `!isSparseCsr()` to the COO version)

Why? `Sparse` corresponds to `COO` layout and `SparseCsr` corresponds to `CSR` layout. The different layouts should not infer with each other, IIUC.

bhosmer commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -8,12 +8,84 @@
+
+} // end anonymous namespace
+
+void _validate_sparse_csr_tensor_args(const Tensor& crow_indices, const Tensor& col_indices, const Tensor& values, IntArrayRef size) {
```
So excellent to have the invariants collected together into one user-callable function - in the spirit of which, wouldn't we also want to check the type and device invariants (currently inlined into `set_member_tensors`) here? 

It's true that we'd want to double-checking these when a call to `_validate_sparse_csr_tensor_args()` is followed by a call to `_sparse_csr_tensor_unsafe()` (which calls `set_member_tensors()` currently), but that seems doable by factoring. 

(Aside: chunking up these invariant checks into helpers that followed the top-level structure of https://github.com/pearu/pearu.github.io/blob/pearu-csr_invariants/csr_tensor_invariants.md - check_type_invariants, etc. - might make this kind of factoring more convenient, plus it'd be pretty! but maybe overkill. just a thought 😁 )

aocsa replied:
Yeap I think the same, It'd be little overkill   :grin:

bhosmer replied:
ahaha ok but so just to be clear, it was the factoring into little helpers that seemed to me like overkill - simply adding the device and type checks to `_validate_sparse_csr_tensor_args()` (and avoiding double checking on construction) doesn't seem gratuitous, it just cleans up the API by making the user-callable invariant checker completely self-contained.

Totally understand your time constraints though - this could certainly be done in a followup by someone else. But I want to understand if there's a *design* reason not to do it that I'm missing.

Re accepting, given that I'm relatively new to the PR and its prior discussion I'll leave final signoff to @ezyang and/or @pearu, but this LGTM - also wanted to note how much @pearu's enumeration of invariants facilitates reasoning about this logic, thanks @pearu 😁

aocsa replied:
Thanks, @bhosmer, now I understand better what u mean. I agree with the idea to have a user-callable invariant checker completely self-contained. Let me update this part.

aocsa replied:
Thanks, @bhosmer, I really appreciate your feedback, and now I understand better what u mean. I agree with the idea to have a user-callable invariant checker completely self-contained. 

So based on this, I can move most of the checks,  except:
 
- `1.4 values.dtype == dtype`
https://github.com/pytorch/pytorch/blob/33c11299f94e5114ec18b0f1809c8c9ce0e763bb/aten/src/ATen/SparseCsrTensorImpl.cpp#L93

-  `4.4 values.device == device`
https://github.com/pytorch/pytorch/blob/33c11299f94e5114ec18b0f1809c8c9ce0e763bb/aten/src/ATen/SparseCsrTensorImpl.cpp#L116

This is because `dtype` and `device` are member variables of class `SparseCsrTensorImpl` and both are not part of the input parameters for `_validate_sparse_csr_tensor_args`.  

IMO having these checks separated in two functions `set_member_tensors` (condition 1.4 and 4.4)  `_validate_sparse_csr_tensor_args` (the rest of 1x and 4x) is not a good idea. 

In the other hand, these type invariants and device invariants checks only make sense when you are creating a new CSR sparse tensor because you are providing the preferred `dtype` and `device` for the tensor construction.   

So what do you think? Looking forward to your comments. cc @pearu

bhosmer replied:
@aocsa ah good point that the invariant situation is slightly more subtle - starting from your analysis, I think there are 
- the self-contained arg invariants, which can be checked in `_validate_sparse_csr_tensor_args_`, and 
- the invariants defined _with respect to the containing tensor_, which can be checked in `set_member_tensors`.

Looking over [the list](https://github.com/pearu/pearu.github.io/blob/main/csr_tensor_invariants.md), I think the latter includes (as you say) 1.4 - values.dtype needs to match the tensor's dtype - but also all of 4 (everybody's device needs to match the tensor's device). 

Note: I think 4.* can be usefully split into 
- a check in `_validate_sparse_csr_tensor_args_` that all subtensors have the same device, and 
- a check of just one of the subtensors against the tensor device in `set_member_tensors`.

pearu replied:
I agree with @bhosmer , checking the devices of all member tensors in `set_member_tensors` is suboptimal.

In fact, new_csr_tensor checks that `device` is `CPU` or `CUDA`. So, checking the devices of all member tensors can be moved to `_validate_sparse_csr_tensor_args`.

pearu replied:
In fact, the same pattern would apply to `dtype`: 1.1 and 1.2 can be moved to `_validate_sparse_csr_tensor_args`.

Is there any reason against moving 1.3 and 1.4 to `_validate_sparse_csr_tensor_args` as well?

aocsa replied:
Sure, if device is restricted to CPU/CUDA  all checking the devices of all member tensors can be moved to _validate_sparse_csr_tensor_args. About dtype, I'm still thinking 1.4 should be at  `set_member_tensors`.  Unlike preferred device (2 options)  preferred dtype for the tensor construction can be  float32 (default), or float64, or int8, ..., or int64. When there is not a dtype supported, that case is caught  in the type dispatcher.

aocsa committed:
rebased/addressed review comments - 3 on "CUDA support in the CSR layout: constructors"


closes gh-56485 , #57211

Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

aocsa commented:
Thanks @bhosmer, @pearu, I rebased this PR, polished a few things based on latest feedback.  As a gentle reminder, this PR is still waiting for a final review  before merging, and I won't be able to work on this after this month.

As a quick note: I added one PR before `CUDA addmm/matvec` #57403  with a fix for CPU version of  `addmm`  op. So, now besides only `matvec` op other ops derived from addmm are supported. cc @pearu, @ezyang, @mruberry

pearu reviewed (commented):
@bhosmer , thanks for bringing up the API design.

Let me summarize the current state: to construct a CSR tensor, we have two methods, safe and unsafe. The safe method first validates the constructor inputs against the layout invariants (crow_indices, col_indices, values, size) and then constructs the tensor using the unsafe method. The unsafe method basically constructs the tensor instance and sets the tensor members `crow_indices_`, `col_indices_`, `values_`, and finally, sets `sizes_and_strides`.

In addition to the above, `_validate_sparse_csr_tensor_args` is exposed as a private function (although it is private, it is still exposed..). It is not clear to me why: when a user or something uses unsafe construction, I would assume that the  `_validate_sparse_csr_tensor_args` was skipped for a reason. If not, the user or something should have used the safe method in the first place. So, I feel that there is some redundancy in the API when exposing all the three tools: safe, unsafe, and validate functions. Or in other words, exposing either "safe and unsafe" or "unsafe and validate" should provide a sufficient and minimal API, but "safe and unsafe and validate" would be redundant as it would enable doing (unreasonable) things like "validate and then use safe to construct".

pearu commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -33,100 +105,90 @@ SparseCsrTensor new_csr_tensor(const TensorOptions& options) {
+
   TensorOptions options = TensorOptions().dtype(dtype).layout(layout).device(device).pinned_memory(pin_memory);
+
+  TORCH_CHECK_NOT_IMPLEMENTED(
+    options.device().type() == kCPU || options.device().type() == kCUDA,
+     "Could not run '", "sparse_csr_tensor", "' from the '", options.device(), "' device.)");
```
This check is redundant as `new_csr_tensor()` call does the same. Remove.

aocsa replied:
Ok, I moved TORCH_CHECK_NOT_IMPLEMENTED to `new_csr_tensor` function.  Just for context TORCH_CHECK_NOT_IMPLEMENTED is important here, because when the `device` parameter provided in the constructor is  `meta` the tests can be skipped.

pearu commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -33,100 +105,90 @@ SparseCsrTensor new_csr_tensor(const TensorOptions& options) {
+    options.device().type() == kCPU || options.device().type() == kCUDA,
+     "Could not run '", "sparse_csr_tensor", "' from the '", options.device(), "' device.)");
+
   TORCH_CHECK(
       options.layout() == kSparseCsr,
       "expected sparse CSR layout, but got layout ",
       options.layout());
```
This check is redundant as `new_csr_tensor()` call does the same. Remove.

pearu reviewed (changes requested):
I have updated https://github.com/pearu/pearu.github.io/blob/main/csr_tensor_invariants.md to reflect the current state of this PR.

There exist two redundant checks. Otherwise, LGTM.

aocsa committed:
rebased/last feedback addressed on "CUDA support in the CSR layout: constructors"


closes gh-56485 , #57211

Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

aocsa commented:
> Thanks @bhosmer, @pearu, I rebased this PR, polished a few things based on latest feedback. As a gentle reminder, this PR is still waiting for a final review before merging, and I won't be able to work on this after this month.
> 
> As a quick note: I added one PR before `CUDA addmm/matvec`,  PR #58622  with a fix for CPU version of `addmm` op. So, now besides only `matvec` op other ops derived from addmm are supported. cc @pearu, @ezyang
> > -- 
> 
> I have updated https://github.com/pearu/pearu.github.io/blob/main/csr_tensor_invariants.md to reflect the current state of this PR.
> 
> There exist two redundant checks. Otherwise, LGTM.

Thanks @pearu, I rebased this PR  and addressed latest feedback. Let me know if anything needs to be taken care of from my end to merge this into PR.  cc @ezyang

pearu commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -613,6 +613,23 @@ Tensor sparse_csr_tensor_ctor(c10::DispatchKey dispatch_key, at::ScalarType scal
+    // Otherwise subsequent Python C API calls might return bogus values.
+    // See https://github.com/pytorch/pytorch/issues/58520 for more details
+    auto rc = PyObject_GetAttrString(o, attr_name);
+    if (!rc) {
```
Notice that you have to check if the exception is really an AttributeError, at least. If it is not, the exception should be raised.
For instance, say, `o` is an instance of the following class:
```
class A:
    @property
    def dtype(self):
        raise NotImplementedError("Say something")
```
in which case the error should not be cleared.

So, use the following logic:
```
if (!rc) {
    if (!PyErr_ExceptionMatches(PyExc_AttributeError)) {
        throw python_error();
    }
    // Warning: a wrong attribute error may be suppressed here
    PyErr_Clear();
}
```

pearu replied:
Ignore the above comment if it is not a part of this PR. Nevertheless, the current behaviour must be fixed at some point.

pearu replied:
Resolving via https://github.com/pytorch/pytorch/issues/58735

aocsa replied:
Thanks Pearu, I will add this logic too.

pearu reviewed (changes requested):
Thanks, @aocsa ! The invariant checking part looks great to me.

I have a nit regarding `safe_get_attr_string`: currently, it clears all errors, not just attribute errors, which is wrong.

bhosmer commented:
Hi @pearu - yeah, you're right that the safe/unsafe/validate trio of methods isn't strictly minimal for just the safe/unsafe creation of a single CSR tensor from a given set of args. My impression is that the motivation for the `safe(args) = unsafe(validate(args))` API structure is to permit the use of validated args in other patterns - say, to create multiple tensors without arg revalidation - but I don't have direct experience of concrete use-cases, it just seems like a reasonable open-world factoring.

pearu commented:
> Hi @pearu - yeah, you're right that the safe/unsafe/validate trio of methods isn't strictly minimal for just the safe/unsafe creation of a single CSR tensor from a given set of args. My impression is that the motivation for the `safe(args) = unsafe(validate(args))` API structure is to permit the use of validated args in other patterns - say, to create multiple tensors without arg revalidation - but I don't have direct experience of concrete use-cases, it just seems like a reasonable open-world factoring.

@bhosmer yes, your example on allowing the creation of multiple tensors makes sense. So, I am fine with the trio.

pearu reviewed (approved):
LGTM

aocsa committed:
run CI/last feedback addressed on "CUDA support in the CSR layout: constructors"


closes gh-56485  and  gh-57211

Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

aocsa committed:
rebased needed on "CUDA support in the CSR layout: constructors"


closes gh-56485  and  gh-57211

Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

aocsa committed:
rebased needed : on "CUDA support in the CSR layout: constructors"


closes gh-56485  and  gh-57211

Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

aocsa committed:
rebased needed after csr addmm fix is merged on "CUDA support in the CSR layout: constructors"


closes gh-56485, gh-57211 and  gh-58762

Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

aocsa commented:
I rebased this stack PR on top of the latest changes. Let's see what CI say but I think this PR is ready for merging. cc @malfet

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28112689).

bhosmer commented:
Hi @aocsa -  looks excellent! LMK if you get a chance to look at the test failures, I kicked off a rerun of `pytorch_xla_linux_bionic_py3_6_clang9_test` but haven't looked at the other. Also watching the internal testing FYI.

aocsa committed:
minor fix for to_dense CUDA/HIP test error on "CUDA support in the CSR layout: constructors"


closes gh-56485, gh-57211 and  gh-58762

Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

aocsa committed:
minor fix for to_dense CUDA/HIP test error on "CUDA support in the CSR layout: constructors"


closes gh-56485, gh-57211 and  gh-58762

Differential Revision: [D28112689](https://our.internmc.facebook.com/intern/diff/D28112689)

[ghstack-poisoned]

aocsa commented:
> Hi @aocsa - looks excellent! LMK if you get a chance to look at the test failures, I kicked off a rerun of `pytorch_xla_linux_bionic_py3_6_clang9_test` but haven't looked at the other. Also watching the internal testing FYI.

Thanks @bhosmer, I figure out the error,  tests related to `csr.to_dense` with CUDA should be enabled in next PR in the stack and not here. Hope this time everything is OK.

bhosmer commented:
@aocsa awesome, thank you for doing the legwork! Still looking good with a few left as I push away from the keyboard, will catch up with it in the AM 😁

aocsa commented:
I think this stack PR is ready to be merged.  CI looks good now with all green 💚.   @bhosmer , @ezyang

bhosmer commented:
@aocsa fyi flaky internal land is why this isn't merged already but I'll be babying it along till it's in 😁

albanD commented:
Sorry, had to revert as this broken a few cuda builds on master like: https://app.circleci.com/pipelines/github/pytorch/pytorch/327701/workflows/92ba24a2-a881-4a6b-9712-9ccbd35e265c/jobs/13731409 and https://app.circleci.com/pipelines/github/pytorch/pytorch/327701/workflows/92ba24a2-a881-4a6b-9712-9ccbd35e265c/jobs/13731498

aocsa commented:
> Sorry, had to revert as this broken a few cuda builds on master like: https://app.circleci.com/pipelines/github/pytorch/pytorch/327701/workflows/92ba24a2-a881-4a6b-9712-9ccbd35e265c/jobs/13731409 and https://app.circleci.com/pipelines/github/pytorch/pytorch/327701/workflows/92ba24a2-a881-4a6b-9712-9ccbd35e265c/jobs/13731498

That's weird, these tests were skipped 
https://github.com/pytorch/pytorch/blob/03b6a0bc8aee90d6e40422d97755616f0d17c4d5/test/test_sparse_csr.py#L282
https://github.com/pytorch/pytorch/blob/03b6a0bc8aee90d6e40422d97755616f0d17c4d5/test/test_sparse_csr.py#L404

Due to the implementation for these methods should be enabled in the next PR in this stack. I Will reopen this PR.

pearu commented:
@aocsa , use `@onlyCPU` on these failing test functions?

aocsa commented:
> @aocsa , use `@onlyCPU` on these failing test functions?

Yeap I will use `@onlyCPU`  on these test functions.

albanD commented:
Note that you can add the ci/master label to run all the master-only tests.
You can use that if you need to check that master-only tests all pass.

aocsa commented:
My stack is poisoned after it was reverted, so I am closing this PR.  
I'm create a new PR here https://github.com/pytorch/pytorch/pull/59004
Hope this time works :+1: :pray:

bhosmer commented:
new new PR: #59010
